src/vcpu.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright 2018-2019 CrowdStrike, Inc.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2018 Cloudbase Solutions Srl
+// SPDX-License-Identifier: Apache-2.0
+//
+// Portions Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the THIRD-PARTY file.
+
+//! Trait to access the functionality of a virtual CPU (vCPU).
+
+use std::{io, result};
+
+///
+/// Generic types used in the virtual CPU trait definition and exported for
+/// public consumption.
+///
+/// These types use kvm_bindings under the hood, as they are not necessarily KVM-
+/// specific, but rather generic x86/x86/ARM structures.  Generic naming makes
+/// them more intuitive for consumption by non-KVM VMMs.
+///
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub use x86_64::{
+    DescriptorTable, FpuState, MsrEntries, MsrEntry, SegmentRegister, SpecialRegisters,
+    StandardRegisters,
+};
+
+///
+/// Generic types used in the virtual CPU trait definition and exported for
+/// public consumption.
+///
+/// These types are explicitly defined in x86_64
+///
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub use x86_64::{CpuId, LapicState};
+
+///
+/// Reasons for vCPU exits.
+///
+/// Exit reasons are drawn a superset of exit reasons for various VMMs.
+///
```

Comment:
I would have preferred to implement `VcpuExit` as an Associated Type (like `RunContextType`), and have each implementation of the Vcpu trait be responsible for defining their own VcpuExit enums, so that we didn't have to guess on the superset and/or update it for changes in the future. However, the kvm-ioctl/Firecracker/CroSVM code all rely on storing ioport and mmio data in the VcpuExit itself, which requires an explicit lifetime annotation on the VcpuExit:

```
pub enum VcpuExit<'a> {
    IoOut(u16 /* port */, &'a [u8] /* data */),
    IoIn(u16 /* port */, &'a mut [u8] /* data */),
    MmioRead(u64 /* address */, &'a mut [u8]),
    MmioWrite(u64 /* address */, &'a [u8]),
    // …
}
```

The only way I was able to successfully implement the VcpuExit as an associated type requires explicit lifetime annotations everywhere: In the trait definition, trait implementations, run function. And, most problematic--each explicit lifetime annotation must be carried over as a generic thereafter.

Example: 
```
// Trait definition
pub trait Vcpu<'a> {
    type RunContextType;
    type VcpuExitType;
    fn run(&'a self) -> Result<Self::VcpuExitType>;

// Trait implementation
impl<'a> Vcpu<'a> for VcpuFd {
    type RunContextType = kvm_bindings::kvm_run;
    type VcpuExitType = VcpuExit<'a>;

     fn run(&'a self) -> Result<VcpuExit<'a>> { … }

// Every future use of the trait as a generic requires the explicit annotation too
pub fn setup_fpu<a', T: Vcpu<'a>>(vcpu: &T) -> Result<()>  {
    vcpu.set_fpu(…)
}
```

IMO, this doesn't make the use of an Associated Type worth it.  As far as I can tell, generic associated types is an open Rust RFC: https://github.com/rust-lang/rust/issues/44265, so it might be doable in the future, but not yet.

More experienced Rust-developers than I: Please let me know if there's a better way to abstract this.

Reply:
We encountered the same issue when designing the vm-memory crate. And the current solution for vm-memory is to carry on the lifetime parameter, and it's really annoying:(